File: ShopNowAndroid/app/src/main/java/com/example/shopnow/data/FakeBackend.kt
package com.example.shopnow.data

import com.example.shopnow.R
import kotlinx.coroutines.delay

class FakeBackend(
    private val networkDelayMs: Long = 2500L
) {
    private var fetchCount: Int = 0

    suspend fun fetchProducts(): List<Product> {
        delay(networkDelayMs)
        fetchCount++

        val base = listOf(
            Product("1","Neon Green Sneakers","€89.00","Lightweight neon green sneakers with great comfort.", R.drawable.shoes1),
            Product("2","Red Sneakers","€89.00","Lightweight red sneakers with great comfort.", R.drawable.shoes2),
            Product("3","Black T-Shirt","€19.00","Soft and premium T-Shirt made from unicorn hair.", R.drawable.t_shirt),
            Product("4","Navi Shirt","€199.00","Probably a navy Shirt", R.drawable.shirt),
        )

        val extra = Product(
            id = "r$fetchCount",
            name = "Promotional refreshable item #$fetchCount",
            price = "€${9 + fetchCount}.00",
            description = "This item appears after refresh to prove refresh works.",
            imageRes = R.drawable.cap
        )

        return if (fetchCount == 1) base else (listOf(extra) + base)
    }
}

File: ShopNowIOS/ShopNowIOS/FakeBackend.swift
import Foundation
import os

actor FakeBackend {

    private let poiLog = OSLog(subsystem: Bundle.main.bundleIdentifier ?? "ShopNowIOS",
                               category: "PointsOfInterest")

    private let delayNanoseconds: UInt64 = 2_500_000_000
    private var fetchCount: Int = 0

    func fetchProducts() async throws -> [Product] {
        // BEGIN signpost
        os_signpost(.begin, log: poiLog, name: "SHOPNOW_FETCH_PRODUCTS")

        defer {
            // END signpost
            os_signpost(.end, log: poiLog, name: "SHOPNOW_FETCH_PRODUCTS")
        }

        try await Task.sleep(nanoseconds: delayNanoseconds)
        fetchCount += 1

        let base = [
            Product(id: "1", name: "Neon Green Sneakers", price: "€89.00",
                    description: "Lightweight neon green sneakers with great comfort.", imageName: "shoes1"),
            Product(id: "2", name: "Red Sneakers", price: "€89.00",
                    description: "Lightweight red sneakers with great comfort.", imageName: "shoes2"),
            Product(id: "3", name: "Black T-Shirt", price: "€19.00",
                    description: "Soft and premium T-Shirt made from unicorn hair.", imageName: "t_shirt"),
            Product(id: "4", name: "Navy Shirt", price: "€199.00",
                    description: "Probably a navy shirt.", imageName: "shirt")
        ]

        let extra = Product(
            id: "r\(fetchCount)",
            name: "Promotional refreshable item #\(fetchCount)",
            price: "€\(9 + fetchCount).00",
            description: "This item appears after refresh to prove refresh works.",
            imageName: "cap"
        )

        return fetchCount == 1 ? base : [extra] + base
    }
}
